File: medium/1306_jump_game_iii/solution.cpp
#include <stack>
#include <vector>

class Solution {
public:
  bool canReach(std::vector<int> &arr, int start) {
    const int n = arr.size();

    std::vector<bool> visited(n, false);

    std::stack<int> s;
    s.push(start);

    while (!s.empty()) {
      const int i = s.top();
      s.pop();

      visited[i] = true;

      if (arr[i] == 0)
        return true;

      const int leftI = i + arr[i];
      const int rightI = i - arr[i];

      if (leftI < n && !visited[leftI])
        s.push(leftI);
      if (rightI > -1 && !visited[rightI])
        s.push(rightI);
    }

    return false;
  }
};

File: medium/1306_jump_game_iii/Solution.java
import java.util.ArrayDeque;
import java.util.Deque;

class Solution {
  public boolean canReach(int[] arr, int start) {
    final int n = arr.length;

    final boolean[] visited = new boolean[n];

    final Deque<Integer> s = new ArrayDeque<>();
    s.push(start);

    while (!s.isEmpty()) {
      final int i = s.pop();

      visited[i] = true;

      if (arr[i] == 0) {
        return true;
      }

      final int leftI = i + arr[i];
      final int rightI = i - arr[i];

      if (leftI < n && !visited[leftI]) {
        s.push(leftI);
      }
      if (rightI > -1 && !visited[rightI]) {
        s.push(rightI);
      }
    }

    return false;
  }
}
